ABI checker: diagnose the missing of @available attributes for added ABIs

nkcsgexi opened the pull request:
New ABIs should have an @available attribute to describe the introducing version. This patch teaches the tool to diagnose its missing.

Decls with @_AlwaysEmitIntoClient are excluded from the blaming lists since they are essentially available all the time.

Thanks for the inputs from @lorentey and @jrose-apple !

rdar://51089418

nkcsgexi committed:
swift-api-digester: describing external type declarations as extensions in error messages

External type declarations are synthesized to incorporate members in extensions to types
of external modules. In diagnostics, we should use 'extension' instead of 'struct/class'
for these decls to avoid confusion.

nkcsgexi committed:
ABI checker: diagnose the missing of @available attributes for added ABIs

New ABIs should have an @available attribute to describe the introducing
version. This patch teaches the tool to diagnose its missing.

Decls with @_AlwaysEmitIntoClient are excluded from the blaming lists since
they are essentially available all the time.

rdar://51089418

nkcsgexi commented:
@swift-ci please smoke test

jrose-apple commented on tools/swift-api-digester/swift-api-digester.cpp:
```diff
@@ -1003,6 +1003,13 @@ class PrunePass : public MatchedNodeListener, public SDKTreeDiffPass {
+          // Diagnose the missing of @available attributes.
+          // Decls with @_alwaysEmitIntoClient aren't required to have an
+          // @available attribute.
+          if (D->getIntroducingVersion().empty() &&
```
Swift availability doesn't count as OS availability.

nkcsgexi replied:
Will Fix!

jrose-apple commented on tools/swift-api-digester/swift-api-digester.cpp:
```diff
@@ -1003,6 +1003,13 @@ class PrunePass : public MatchedNodeListener, public SDKTreeDiffPass {
+          // @available attribute.
+          if (D->getIntroducingVersion().empty() &&
+              !D->hasDeclAttribute(DeclAttrKind::DAK_AlwaysEmitIntoClient)) {
+            D->emitDiag(diag::new_decl_without_intro);
```
Can this be an optional diag? It's only relevant for APIs shipped in Apple OSs.

nkcsgexi replied:
Is this really necessary? Most frameworks using this tool are in such case.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi committed:
swift-api-digester: add a flag to disable OS related diagnostics

For frameworks not shipping with OSs, the users should specify this flag
to avoid diagnosing some changes, such as the missing of OS availability
attributes.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi committed:
test/abi-checker: run stdlib ABI stability test only when assertions are enabled

We started to diagnose new APIs such as the missing of @available attributes. This diverges
the expected change lists for builds with/without assertions enabled because some stdlib
symbols are included only when asserts are enabled. We start to require
asserts here to make sure the test doesn't fail on no-assertion settings.

nkcsgexi commented:
@swift-ci please smoke test

jrose-apple commented on test/api-digester/stability-stdlib-abi.swift:
```diff
@@ -1,4 +1,5 @@
 // REQUIRES: OS=macosx
+// REQUIRES: asserts
```
This seems very dangerous. Can we have a with-asserts and a without-asserts variant, and say that the dump has to match one of them?

nkcsgexi replied:
It sounds like a good idea. I'll add another test requires no asserts

jrose-apple replied:
It'd be easiest to keep them in sync if we assume the with-asserts variant is a superset. Then it can just use the `cat` of the two files as the expected changes.

nkcsgexi replied:
Just split the expected changes to two lists: one without asserts and the other the additional blames when asserts are enabled. The asserts-enabled test will diff against the combination of both.

nkcsgexi committed:
test/abi-checker: add a stdlib ABI stability test when asserts are disabled

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test
